### PR TITLE
Update node types declaration to latest v16

### PR DESCRIPTION
**Pull request by swbsi**

### CODE REVIEW

File: core/common/src/rpc/web/WebAppRpcProtocol.ts [hunk 1]
```diff
@@ -36,7 +36,7 @@ export interface HttpServerRequest extends Readable {
   statusCode?: number;
   statusMessage?: string;
   socket: any;
-  destroy(error?: Error): void;
+  destroy(error?: Error): any;
```
Comment (aruniverse) [inline core/common/src/rpc/web/WebAppRpcProtocol.ts:39]: whats the reason for this change?

File: core/backend/package.json [hunk 2]
```diff
@@ -68,7 +68,7 @@
     "@types/fs-extra": "^4.0.7",
     "@types/mocha": "^8.2.2",
     "@types/multiparty": "^0.0.31",
-    "@types/node": "16.11.7",
+    "@types/node": "18.7.18",
```
Comment (aruniverse) [inline core/backend/package.json:71]: do we want to be pinning it to an exact version? or should we be using ^ to pull the latest types?

**Review by aruniverse — COMMENTED** [review]
wonder if we really should have a dep on @types/node for frontend pkgs
could you update rush.json to support node 18?

**Comment by ben-polinsky** [comment]
@calebmshafer @aruniverse I have this assigned to me https://github.com/iTwin/itwinjs-core/issues/4135 and have been working on it. Is there a way to prevent duplication of work so time isn't wasted?

File: core/common/src/rpc/web/WebAppRpcProtocol.ts (hunk 1 above)
Comment (pmconne) [reply core/common/src/rpc/web/WebAppRpcProtocol.ts:39]: #3947

Comment (swbsi) [reply core/common/src/rpc/web/WebAppRpcProtocol.ts:39]: Will we be supporting both node 16 and 18? If so, the return type differs (void on 16 and this on 18).
Do you think I should do it differently?

**Comment by swbsi** [comment]
> @calebmshafer @aruniverse I have this assigned to me #4135 and have been working on it. Is there a way to prevent duplication of work so time isn't wasted?

I wasn't planning to work on the broader v18 project. Caleb just asked me to get this one through. I got involved through https://github.com/iTwin/itwinjs-core/issues/3947.

**Comment by swbsi** [comment]
> wonder if we really should have a dep on @types/node for frontend pkgs could you update rush.json to support node 18?

I think ideally not in the frontend, although perhaps my RPC stuff in common is forcing it for now?

File: core/backend/package.json (hunk 2 above)
Comment (swbsi) [reply core/backend/package.json:71]: > do we want to be pinning it to an exact version? or should we be using ^ to pull the latest types?

Wouldn't this silently pull us up to node versions that we aren't officially supporting yet, though?

File: core/common/src/rpc/web/WebAppRpcProtocol.ts (hunk 1 above)
Comment (ben-polinsky) [reply core/common/src/rpc/web/WebAppRpcProtocol.ts:39]: 18 returns `this` for destroy? I added an overload:

```ts
  destroy(error?: Error): this;
  destroy(error?: Error): void;
```
But could also use `this|void`? What do you think?

Comment (swbsi) [reply core/common/src/rpc/web/WebAppRpcProtocol.ts:39]: I tried this|void, which was a compiler error.
I didn't think of an overload...that seems to work. I will push that.

Comment (aruniverse) [reply core/common/src/rpc/web/WebAppRpcProtocol.ts]: [node@14](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/v14/stream.d.ts#L48)
[node@16](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/v16/stream.d.ts#L411)
[node@18](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/stream.d.ts#L434)

Looks like the return type has been `this` for quite some time now, making me question the need to update to 18 with this pr.

**Comment by swbsi** [comment]
> wonder if we really should have a dep on @types/node for frontend pkgs could you update rush.json to support node 18?

Should I update rush.json to support 18 in this PR or let Ben do it in the larger "full support" PR?

File: core/common/src/rpc/web/WebAppRpcProtocol.ts (hunk 1 above)
Comment (swbsi) [reply core/common/src/rpc/web/WebAppRpcProtocol.ts]: > [node@14](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/v14/stream.d.ts#L48) [node@16](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/v16/stream.d.ts#L411) [node@18](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/stream.d.ts#L434)
> 
> Looks like the return type has been `this` for quite some time now, making me question the need to update to 18 with this pr.

I guess we're just running an "incorrect" version of @node/types, then (16.11.7)?

I guess my PR here is just getting us to a better version of @node/types and definitely has nothing to do with the broader node18 update, then.

Comment (swbsi) [reply core/common/src/rpc/web/WebAppRpcProtocol.ts]: > The type used to be `void` back in [node@10](https://unpkg.com/browse/@types/node@10.17.60/stream.d.ts), which we don't currently support anymore anyway.

I will take out my overload for void return then...

Comment (aruniverse) [reply core/common/src/rpc/web/WebAppRpcProtocol.ts]: try changing the types ver from `"16.11.7"` to `"^16.11.7", then run `rush update --full`. Afterwards go back to WebAppRpcProtocol, you might need to restart the TS Server (open the vscode command palette and run TypeScript Restart TS Server). You will see theres a warning about the incorrect return type now

image

image

File: core/backend/package.json (hunk 2 above)
Comment (aruniverse) [reply core/backend/package.json:71]: Using ^ should be fine, itll only pull the latest minor/patch in this major, and Node follows semver better than us, so we can expect no breaking changes

File: core/common/src/rpc/web/WebAppRpcProtocol.ts (hunk 1 above)
Comment (aruniverse) [reply core/common/src/rpc/web/WebAppRpcProtocol.ts]: if using the carrot doesnt pull the latest 16.x release, then just change it to `16.11.59`

**Review by mdastous-bentley — APPROVED** [review]

File: core/backend/package.json (hunk 2 above)
Comment (swbsi) [reply core/backend/package.json:71]: Done.

File: core/common/src/rpc/web/WebAppRpcProtocol.ts (hunk 1 above)
Comment (swbsi) [reply core/common/src/rpc/web/WebAppRpcProtocol.ts]: Yes, I got the error also. Guess we should have had the caret all along...

Comment (aruniverse) [reply core/common/src/rpc/web/WebAppRpcProtocol.ts]: so what im getting at is we actually dont need to bump to ^18, just bump to ^16 in this pr and change the return type.

@ben-polinsky will work on actually supporting node18, and will include the types/node!18 in his pr

Comment (swbsi) [reply core/common/src/rpc/web/WebAppRpcProtocol.ts]: > so what im getting at is we actually dont need to bump to ^18, just bump to ^16 in this pr and change the return type.
> 
> @ben-polinsky will work on actually supporting node18, and will include the types/node!18 in his pr

Will do.

Comment (swbsi) [reply core/common/src/rpc/web/WebAppRpcProtocol.ts]: Done. That hat didn't seem to get rush update to move up, so I hard coded 11.59.

**Review by aruniverse — APPROVED** [review]

**Review by ben-polinsky — APPROVED** [review]

**Review by jchick-bentley — APPROVED** [review]

**Review by dassaf4 — APPROVED** [review]

**Review by pmconne — APPROVED** [review]

**Review by paulius-valiunas — APPROVED** [review]

**Review by diegoalexdiaz — APPROVED** [review]

**Review by ColinKerr — APPROVED** [review]

**Review by calebmshafer — APPROVED** [review]

**Comment by calebmshafer** [comment]
@swbsi mind fixing the merge conflicts?

@aruniverse after he does that, feel free to force merge.

**Review by MichaelBelousov — APPROVED** [review]

**Review by NancyMcCallB — APPROVED** [review]

**Comment by swbsi** [comment]
> @swbsi mind fixing the merge conflicts?
> 
> @aruniverse after he does that, feel free to force merge.

Done.

**Review by grigasp — APPROVED** [review]

**Pull request merged**